src/main/java/com/bpodgursky/jbool_expressions/NExpression.java

```diff
@@ -12,14 +12,19 @@
 
   public final Expression<K>[] expressions;
 
-  protected NExpression(List<? extends Expression<K>> expressions){
+  protected NExpression(List<? extends Expression<K>> expressions, long hashBase){
     if(expressions.isEmpty()){
       throw new IllegalArgumentException("Arguments length 0!");
     }
 
     this.expressions = expressions.toArray(ExprUtil.<K>expr(0));
-    this.hash = expressions.stream().mapToInt(Expression::hashCode)
-                           .reduce(1, (s, c) -> s * c);
+    this.hash = hashBase;
+    for(Expression e: expressions)
+    {
+      this.hash = this.hash * 31;
+      this.hash = this.hash + e.hash;
+    }
+
     Arrays.sort(this.expressions);
```

Comment:
will need the arrays.sort call above the hash code I think

also we should probably make the hash value final

Reply:
We can't make `hash` final in `Expression` as we'd have to initialise it there. We could have a final `hash` in `NExpression` though?

Moving the sort above the hashing makes no difference to comparing `(A|B)` with `(B|A)` sadly.